sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -172,3 +172,93 @@ def test_binning_train_validation_are_separated():
+    is True.
+
+    """
+    estimator = GradientBoosting(max_iter=50, warm_start=True)
+    estimator.fit(X, y)
+    estimator.set_params(max_iter=25)
+    err_msg = ('max_iter=25 must be larger than or equal to n_iter_=50 '
+               'when warm_start==True')
+    with pytest.raises(ValueError, match=err_msg):
+        estimator.fit(X, y)
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_yields_identical_results(GradientBoosting, X, y):
+    """Test that two fits with warm start yield the same results as one fit.
+
+    Make sure that fitting 50 iterations and then 25 with warm start is
+    equivalent to fitting 75 iterations.
+
+    """
+    rng = 42
+    gb_warm_start = GradientBoosting(
+        n_iter_no_change=100, max_iter=50, random_state=rng, warm_start=True
+    )
+    gb_warm_start.fit(X, y).set_params(max_iter=75).fit(X, y)
+
+    gb_no_warm_start = GradientBoosting(
+        n_iter_no_change=100, max_iter=75, random_state=rng, warm_start=False
+    )
+    gb_no_warm_start.fit(X, y)
+
+    # Check identical nodes for each tree
+    for (pred_ith_warm, pred_ith_no_warm) in zip(gb_warm_start._predictors,
+                                                 gb_no_warm_start._predictors):
+        for (predictor_warm, predictor_no_warm) in zip(pred_ith_warm,
+                                                       pred_ith_no_warm):
+            np.testing.assert_array_equal(
+                predictor_warm.nodes,
+                predictor_no_warm.nodes
+            )
+
+    # Check identical predictions
+    np.testing.assert_allclose(gb_warm_start.predict(X),
+                               gb_no_warm_start.predict(X))
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_max_depth(GradientBoosting, X, y):
+    """I did not understand what you expect."""
+    pass
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_early_stopping(GradientBoosting, X, y):
+    """Test that early stopping works as expected with warm starting.
+
+    Make sure that early stopping occurs after a small number of iterations
+    when fitting a second time with warm starting.
+
+    """
+    n_iter_no_change = 5
+    gb = GradientBoosting(
+        n_iter_no_change=n_iter_no_change, max_iter=10000,
+        random_state=42, warm_start=True, tol=1e-3
+    )
+    gb.fit(X, y)
+    n_iter_first_fit = gb.n_iter_
+    gb.fit(X, y)
+    n_iter_second_fit = gb.n_iter_
+    assert n_iter_second_fit - n_iter_first_fit < 2 * n_iter_no_change
```

Comment:
out of curiosity what's the actual value of `n_iter_second_fit - n_iter_first_fit` here?

Reply:
I changed the value of `tol` (`tol=1e-3` instead of the default value `tol=1e-7`) because it fails otherwise for the regression...:
* `tol=1e-3` : classification (6 -> 7), regression (103 -> 105)
* `tol=1e-7` : classification (6 -> 7), regression (105 -> 141)

It feels a bit tricky to change the default value of `tol`, but when the sample size is small (100 by default in `make_classification` and `make_regression`), the validation set is very small (10 by default) and the performance evaluation is a bit noisy.

Reply:
Sounds good. Then please remove the factor 2.